core/src/main/java/org/apache/accumulo/core/zookeeper/ZooCache.java

```diff
@@ -59,77 +62,72 @@ public interface ZooCacheWatcher extends Consumer<WatchedEvent> {}
 
   private static final Logger log = LoggerFactory.getLogger(ZooCache.class);
 
-  private final ZCacheWatcher watcher = new ZCacheWatcher();
-  private final Optional<ZooCacheWatcher> externalWatcher;
+  protected volatile NavigableSet<String> watchedPaths =
+      Collections.unmodifiableNavigableSet(new TreeSet<>());
```

Comment:
The lifecycle of this field is confusing, because it is initialized to an empty set here, then in the constructor, setupWatchers is called, which sets this again to the set passed into the constructor. Then, when ZK connection changes are detected, setupWatchers is called again, using the contents of this set as input, which again clobbers this field with the same set.

Instead, just make this a final, set it once in the constructor (as a copy of the input set), and use it in setupWatchers directly, without modifying it, and without the need to pass it as a parameter to setupWatchers (because it's already an instance member and setupWatchers is a non-static instance method).

Reply:
Yeah, it was confusing before when ZooSession was keeping track of the persistent watchers. That has all been removed now so this is much cleaner now. Changed in a481f9b.